datafusion/physical-plan/src/aggregates/group_values/row_like.rs

```diff
@@ -0,0 +1,310 @@
+                            v.push(Box::new(b) as _)
+                        }
+                        &DataType::UInt64 => {
+                            let b = PrimitiveGroupValueBuilder::<UInt64Type>::default();
+                            v.push(Box::new(b) as _)
+                        }
+                        &DataType::Float32 => {
+                            let b = PrimitiveGroupValueBuilder::<Float32Type>::default();
+                            v.push(Box::new(b) as _)
+                        }
+                        &DataType::Float64 => {
+                            let b = PrimitiveGroupValueBuilder::<Float64Type>::default();
+                            v.push(Box::new(b) as _)
+                        }
+                        &DataType::Date32 => {
+                            let b = PrimitiveGroupValueBuilder::<Date32Type>::default();
+                            v.push(Box::new(b) as _)
+                        }
+                        &DataType::Date64 => {
+                            let b = PrimitiveGroupValueBuilder::<Date64Type>::default();
+                            v.push(Box::new(b) as _)
+                        }
+                        &DataType::Utf8 => {
+                            let b = ByteGroupValueBuilder::<i32>::new(OutputType::Utf8);
+                            v.push(Box::new(b) as _)
+                        }
+                        &DataType::LargeUtf8 => {
+                            let b = ByteGroupValueBuilder::<i64>::new(OutputType::Utf8);
+                            v.push(Box::new(b) as _)
+                        }
+                        dt => todo!("{dt} not impl"),
+                    }
+                }
+                v
+            }
+        };
+
+        // tracks to which group each of the input rows belongs
+        groups.clear();
+
+        // 1.1 Calculate the group keys for the group values
+        let batch_hashes = &mut self.hashes_buffer;
+        batch_hashes.clear();
+        batch_hashes.resize(n_rows, 0);
+        create_hashes(cols, &self.random_state, batch_hashes)?;
+
+        for (row, &target_hash) in batch_hashes.iter().enumerate() {
+            let entry = self.map.get_mut(target_hash, |(exist_hash, group_idx)| {
+                // Somewhat surprisingly, this closure can be called even if the
+                // hash doesn't match, so check the hash first with an integer
+                // comparison first avoid the more expensive comparison with
+                // group value. https://github.com/apache/datafusion/pull/11718
+                if target_hash != *exist_hash {
+                    return false;
+                }
+
+                fn check_row_equal(
+                    array_row: &dyn ArrayRowEq,
+                    lhs_row: usize,
+                    array: &ArrayRef,
+                    rhs_row: usize,
+                ) -> bool {
+                    array_row.equal_to(lhs_row, array, rhs_row)
```

Comment:
Nulls is handled in the `equal_to`, is there any other place we could further optimize?

Reply:
One idea I had is that you could defer actually copying the new rows into group_values so rather than calling the function once for each new group, you could call it once per batch, and it could insert all the new values in one function call

That would save some function call overhead as well as the downcasting of arrays and maybe would vectorize better